ttl/ttlworker/scan.go

```diff
@@ -0,0 +1,276 @@
+	taskMaxErrorRate                = 0.4
+)
+
+type ttlStatistics struct {
+	TotalRows   atomic.Uint64
+	SuccessRows atomic.Uint64
+	ErrorRows   atomic.Uint64
+}
+
+func (s *ttlStatistics) IncTotalRows(cnt int) {
+	s.TotalRows.Add(uint64(cnt))
+}
+
+func (s *ttlStatistics) IncSuccessRows(cnt int) {
+	s.SuccessRows.Add(uint64(cnt))
+}
+
+func (s *ttlStatistics) IncErrorRows(cnt int) {
+	s.ErrorRows.Add(uint64(cnt))
+}
+
+func (s *ttlStatistics) Reset() {
+	s.SuccessRows.Store(0)
+	s.ErrorRows.Store(0)
+	s.TotalRows.Store(0)
+}
+
+type ttlScanTask struct {
+	tbl        *cache.PhysicalTable
+	expire     time.Time
+	rangeStart []types.Datum
+	rangeEnd   []types.Datum
+	statistics *ttlStatistics
+}
+
+type ttlScanTaskExecResult struct {
+	task *ttlScanTask
+	err  error
+}
+
+func (t *ttlScanTask) result(err error) *ttlScanTaskExecResult {
+	return &ttlScanTaskExecResult{task: t, err: err}
+}
+
+func (t *ttlScanTask) getDatumRows(rows []chunk.Row) [][]types.Datum {
+	datums := make([][]types.Datum, len(rows))
+	for i, row := range rows {
+		datums[i] = row.GetDatumRow(t.tbl.KeyColumnTypes)
+	}
+	return datums
+}
+
+func (t *ttlScanTask) doScan(ctx context.Context, delCh chan<- *ttlDeleteTask, sessPool sessionPool) *ttlScanTaskExecResult {
+	rawSess, err := getSession(sessPool)
+	if err != nil {
+		return t.result(err)
+	}
+	defer rawSess.Close()
+
+	sess := newTableSession(rawSess, t.tbl, t.expire)
+	generator, err := sqlbuilder.NewScanQueryGenerator(t.tbl, t.expire, t.rangeStart, t.rangeEnd)
+	if err != nil {
+		return t.result(err)
+	}
+
+	retrySQL := ""
+	retryTimes := 0
+	var lastResult [][]types.Datum
+	for {
+		if err = ctx.Err(); err != nil {
+			return t.result(err)
+		}
+
+		if total := t.statistics.TotalRows.Load(); total > uint64(taskStartCheckErrorRateCnt) {
+			if t.statistics.ErrorRows.Load() > uint64(float64(total)*taskMaxErrorRate) {
+				return t.result(errors.Errorf("error exceeds the limit"))
+			}
+		}
+
+		sql := retrySQL
+		if sql == "" {
+			limit := int(variable.TTLScanBatchSize.Load())
+			if sql, err = generator.NextSQL(lastResult, limit); err != nil {
+				return t.result(err)
+			}
+		}
+
+		if sql == "" {
+			return t.result(nil)
+		}
+
+		rows, retryable, sqlErr := sess.ExecuteSQLWithCheck(ctx, sql)
+		if sqlErr != nil {
+			needRetry := retryable && retryTimes < scanTaskExecuteSQLMaxRetry && ctx.Err() == nil
+			logutil.BgLogger().Error("execute query for ttl scan task failed",
+				zap.String("SQL", sql),
+				zap.Int("retryTimes", retryTimes),
+				zap.Bool("needRetry", needRetry),
+				zap.Error(err),
+			)
+
+			if !needRetry {
+				return t.result(sqlErr)
+			}
+			retrySQL = sql
+			retryTimes++
+			select {
+			case <-ctx.Done():
+				return t.result(ctx.Err())
+			case <-time.After(scanTaskExecuteSQLRetryInterval):
+			}
+			continue
+		}
+
+		retrySQL = ""
+		retryTimes = 0
+		lastResult = t.getDatumRows(rows)
+		if len(rows) == 0 {
+			continue
+		}
+
+		delTask := &ttlDeleteTask{
```

Comment:
Should `delTask` also have some way to send back the execution status? (like the `notifyStateCh` above).

The problem I considered is that the job and manager has no idea how much del tasks will be created, so they don't know how much state they will need to wait. But they actually need to know when all tasks have finished :thinking: (to update state, and migrate all fields from `current_job_*` to `last_job_*`).

A way I could propose is to use a counter to count the number of spawned task of the job, when a task (no matter it's scan or del, and no matter it's error or not) finished, decrease the counter and if the counter arrives 0, update the state and migrate the fields.

See https://github.com/pingcap/tidb/pull/39472/files#diff-ce94329d9478a76725965679f8e00f713c604a630bad1220443582a4a50b8d88R103.

When a scan task want to spawn some del task, it could firstly add the counter (by `task.tracker.Add()`), and when a task is finished, `task.tracker.Done(xxx)` is called (by scan and del worker). :thinking: 

(For this approach, using function call, or a channel with extra goroutine for each job are both fine. I'm not sure which one is better.)

Reply:
I have also considered to track del task by scan worker, and track scan task by job (manager). It seems more complicated :facepalm: .

Reply:
Each task have a ttlStatistics object, I think you can use it to determine whether a all deletions from a scan task is end (by evaluating totalCount == errorCount + successCount).